capsules/extra/src/hc_sr04.rs

```diff
@@ -0,0 +1,158 @@
+    alarm: &'a A,
+    start_time: Cell<u32>,
+    end_time: Cell<u32>,
+    status: Cell<Status>,
+    distance_client: OptionalCell<&'a dyn sensors::DistanceClient>,
+}
+
+impl<'a, A: Alarm<'a>> HcSr04<'a, A> {
+    /// Create a new HC-SR04 driver.
+    pub fn new(
+        trig: &'a dyn kernel::hil::gpio::Pin,
+        echo: &'a dyn kernel::hil::gpio::InterruptPin<'a>,
+        alarm: &'a A,
+    ) -> HcSr04<'a, A> {
+        // Setup and return struct.
+        HcSr04 {
+            trig,
+            echo,
+            alarm,
+            start_time: Cell::new(0),
+            end_time: Cell::new(0),
+            status: Cell::new(Status::Idle),
+            distance_client: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>> Distance<'a> for HcSr04<'a, A> {
+    /// Set the client for distance measurement results.
+    fn set_client(&self, distance_client: &'a dyn DistanceClient) {
+        self.distance_client.set(distance_client);
+    }
+    /// Start a distance measurement.
+    fn read_distance(&self) -> Result<(), ErrorCode> {
+        if self.status.get() == Status::Idle {
+            self.status.set(Status::TriggerPulse);
+            self.trig.set(); // Send the trigger pulse.
+            self.alarm
+                .set_alarm(self.alarm.now(), self.alarm.ticks_from_ms(10));
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+    /// Get the maximum distance the sensor can measure in mm
+    fn get_maximum_distance(&self) -> u32 {
+        // The maximum distance is determined by the maximum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the maximum measurable distance is approximately 4 meters.
+        // Convert this to millimeters.
+        4000
+    }
+    /// Get the minimum distance the sensor can measure in mm.
+    fn get_minimum_distance(&self) -> u32 {
+        // The minimum distance is determined by the minimum pulse width the sensor can detect.
+        // As specified in the datasheet: https://www.handsontec.com/dataspecs/HC-SR04-Ultrasonic.pdf,
+        // the minimum measurable distance is approximately 2 cm.
+        // Convert this to millimeters.
+        20
+    }
+}
+
+impl<'a, A: Alarm<'a>> AlarmClient for HcSr04<'a, A> {
+    /// Handle the alarm event.
+    fn alarm(&self) {
+        match self.status.get() {
+            Status::TriggerPulse => {
+                self.trig.clear(); // Clear the trigger pulse.
+                self.echo.enable_interrupts(InterruptEdge::RisingEdge); // Enable rising edge interrupt on echo pin.
+                self.status.set(Status::EchoStart); // Update status to waiting for echo.
+            }
+            _ => {}
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>> Client for HcSr04<'a, A> {
+    /// Handle the GPIO interrupt.
+    fn fired(&self) {
+        let time = self.alarm.now().into_u32();
```

Comment:
What's guaranteeing that this is in unites of microseconds? Shouldn't it depend on the Alarm Frequency?

Reply:
We need the `end_time` member to calculate the duration of the echo pulse. The time difference between `start_time` and `end_time` gives us the travel time of the sound wave, allowing us to compute the distance.

Reply:
Given that you only use `end_time` in the same function where it's stored in the struct though, it could just be a local variable rather than a struct member, right?